File: 155-min-stack/min-stack.py
class MinStack:

    def __init__(self):
        self.stack=[]
        self.mStack=[]
    def push(self, val: int) -> None:
        mVal=min(val,self.mStack[-1] if self.mStack else val)
        self.stack.append(val)
        self.mStack.append(mVal)
    def pop(self) -> None:
        self.stack.pop()
        self.mStack.pop()
    def top(self) -> int:
        return self.stack[-1]
    def getMin(self) -> int:
        return self.mStack[-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 155-min-stack/min-stack.java
class MinStack {
    Stack<Integer> stack;
    Stack<Integer> minStack;

    public MinStack() {
        this.stack= new Stack<>();
        this.minStack= new Stack<>();
    }
    
    public void push(int val) {
        this.stack.add(val);
        if(this.minStack.isEmpty()||val<this.minStack.peek()){
            this.minStack.add(val);
        }else{
            this.minStack.add(this.minStack.peek());
        }
    }
    
    public void pop() {
        this.stack.pop();
        this.minStack.pop();
    }
    
    public int top() {
        return this.stack.peek();
    }
    
    public int getMin() {
        return this.minStack.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */